packages/flutter/lib/src/widgets/implicit_animations.dart

```diff
@@ -607,3 +608,176 @@ class _AnimatedDefaultTextStyleState extends AnimatedWidgetBaseState<AnimatedDef
+enum CrossFadeState {
+  /// A state where the first child is shown and the second child is completely
+  /// invisible.
+  showFirst,
+  /// A state where the second child is shown and the first child is completely
+  /// invisible.
+  showSecond
+}
+
+/// A widget that cross-fades between two children and animates between their
+/// sizes. The animation is controlled through the [crossFadeState] parameter.
+/// [firstCurve] and [secondCurve] represent the opacity curves of the two
+/// children, and should usually be [Interval] objects. Note that [firstCurve]
+/// is inverted, i.e. it fades out when providing a growing curve like
+/// [Curves.linear]. [sizeCurve] is the curve used to animated between the size
+/// of the fading out child and the size of the fading in child.
+///
+/// In the case where the two children have different sizes, the animation crops
+/// overflowing children during the animation by alining them to the top-left.
+/// This means that the bottom-right size will be cropped until the animation
+/// reaches its end.
+class AnimatedCrossFade extends StatefulWidget {
+  /// Creates a cross fade animation widget.
+  ///
+  /// The [duration] of the animation is the same for all components (fade in,
+  /// fade out, and size), and you can pass [Interval]s instead of [Curve]s in
+  /// order to have finer control.
+  AnimatedCrossFade({
+    Key key,
+    this.firstChild,
+    this.secondChild,
+    this.firstCurve: Curves.linear,
+    this.secondCurve: Curves.linear,
+    this.sizeCurve: Curves.linear,
+    @required this.crossFadeState,
+    @required this.duration
+  }) : super(key: key);
+
+  /// The first child. It fades both in an out depending on [crossFadeState].
+  final Widget firstChild;
+
+  /// The second child. It fades both in an out depending on [crossFadeState].
+  final Widget secondChild;
+
+  /// The state that the cross fade animation is animating towards.
+  final CrossFadeState crossFadeState;
+
+  /// The duration of the whole orchestrated animation.
+  final Duration duration;
+
+  /// The fade curve of the first child. Note that it is inverted, i.e. it
+  /// fades out when providing a growing curve like [Curves.linear].
+  final Curve firstCurve;
+
+  /// The fade curve of the second child.
+  final Curve secondCurve;
+
+  /// The size curve of the widget.
+  final Curve sizeCurve;
+
+  @override
+  _AnimatedCrossFadeState createState() => new _AnimatedCrossFadeState();
+}
+
+class _AnimatedCrossFadeState extends State<AnimatedCrossFade> {
+  _AnimatedCrossFadeState() : super();
+
+  AnimationController _controller;
+  Animation<double> _firstAnimation;
+  Animation<double> _secondAnimation;
+
+  @override
+  void initState() {
+    super.initState();
+    _controller = new AnimationController(duration: config.duration)
+        ..value = 0.0;
+    _firstAnimation = new Tween<double>(
+      begin: 1.0,
+      end: 0.0
+    ).animate(
```

Comment:
I think this would be slightly cleaner if you used `.chain(config.firstCurve).animate(_controller)`.